lib/Sema/TypeCheckStorage.cpp

```diff
@@ -762,38 +812,27 @@ static Expr *buildStorageReference(AccessorDecl *accessor,
     return finish(storageDRE);
   }
 
-  bool isMemberLValue = isLValue;
-  auto propertyWrapperMutability =
-      [&](Decl *decl) -> Optional<std::pair<bool, bool>> {
-    auto var = dyn_cast<VarDecl>(decl);
-    if (!var)
-      return None;
-    auto mut = var->getPropertyWrapperMutability();
-    if (!mut)
-      return None;
-    return std::make_pair(mut->Getter == PropertyWrapperMutability::Mutating,
-                          mut->Setter == PropertyWrapperMutability::Mutating);
-  };
+  // Build self
 
-  // If we're accessing a property wrapper, determine if the
-  // intermediate access requires an lvalue.
-  if (auto mut = propertyWrapperMutability(accessor->getStorage())) {
-    isMemberLValue = mut->first;
-    if (isLValue)
-      isMemberLValue |= mut->second;
+  bool isGetterMutating = storage->isGetterMutating();
+  bool isSetterMutating = storage->isSetterMutating();
+  if (target == TargetImpl::Wrapper || target == TargetImpl::WrapperStorage) {
+    auto var = cast<VarDecl>(accessor->getStorage());
+    if (auto mutability = var->getPropertyWrapperMutability()) {
+      // We consider the storage's mutability too because the wrapped property
+      // might be part of a class, in case of which nothing is mutating.
+      isGetterMutating = (mutability->Getter == PropertyWrapperMutability::Mutating)
+        ? (storage->isGetterMutating() || storage->isSetterMutating())
```

Comment:
This block of code took me a minute to figure out. In the comment above, it might be useful to explain:
1) The storage getter is used to access the wrappedValue/projectedValue in both cases
2) If the getter/setter of the wrappedValue/projectedVar is mutating, you need to consider both the getter and setter of the backing storage because you need both in order to perform the writeback

Reply:
I don't quite understand what you're saying, so let me explain my reasoning.

Here, we're trying to figure out if `self` should be an lvalue or not. `self` should be an lvalue if the current method being synthesised is mutating. If we're using property wrappers, we try to derive that by computing PropertyWrapperMutability. 

I could've written this as [1]:
~~~
isGetterMutating = (mutability->Getter == PropertyWrapperMutability::Mutating);
isSetterMutating = (mutability->Setter == PropertyWrapperMutability::Mutating);
~~~

However, that only applies if `self` was a struct. If `self` was a class, then neither the setter or the getter should be mutating, and in that case, the storage's getter/setter wouldn't be mutating too. So to handle the class scenario, the code looks like this [2]:

~~~
isGetterMutating = (mutability->Getter == PropertyWrapperMutability::Mutating)
  ? (storage->isGetterMutating() || storage->isSetterMutating())
  : storage->isGetterMutating();
isSetterMutating = (mutability->Setter == PropertyWrapperMutability::Mutating)
  ? (storage->isGetterMutating() || storage->isSetterMutating())
  : storage->isGetterMutating();
~~~

Actually, if PropertyWrapperMutability could handle this class scenario internally (i.e. if it could return the correct mutability for classes as well), that would be great. Then we could just use [1].